lib/Scheduler/aco.cpp

```diff
@@ -251,14 +253,59 @@ FUNC_RESULT ACOScheduler::FindSchedule(InstSchedule *schedule_out,
                                        SchedRegion *region) {
   rgn_ = region;
 
+  if(!aco_use_two_pass)
+  {
+    pass = ACOPass::ALL_IN_ONE;
+    return performSchedPass(schedule_out);
+  }
+  else
+  {
+    //in the first pass we schedule for register pressure only
+    //in the second pass we schedule for lower schedule length
+    //but only accept schedules with an RP <= the first passes RP
+    //the first pass result is the initial schedule the second pass
+    pass = ACOPass::FIRST;
+    FUNC_RESULT pass1Res = performSchedPass(schedule_out);
+	if(pass1Res!=RES_SUCCESS) {
+      return pass1Res;
+    }
+    pass = ACOPass::SECOND;
+    setInitialSched(schedule_out);
+    return performSchedPass(schedule_out);
+
+  }
+}
+
+//compares to schedules taking into account which pass
+//we are on and returns whether the old schedule
+//is inferior to the new one and ought to be replaced
+bool ACOScheduler::shouldReplaceSchedule(ACOPass pass, InstSchedule* oldSched,
```

Comment:
This can be static

Reply:
Do we need this function? I don't see a problem with the way we did it before. If we're using two-pass with the settings controlled by the wrapper, the first pass will only have instructions of latency 1. All schedules for the same scheduling region will have the same schedule length but differing RP costs so we only compare RP costs anyways in the first pass. In the second pass, the DDG is re-built to account for latencies so we will compare both schedule length (with stalls) and RP costs. If the RP weight is set high enough, the RP shouldn't ever decrease.

Reply:
The ACO Scheduler is currently called from SchedRegion::FindOptimalSchedule which passes through the DDG that was created in the wrapper to ACO.  If we were to create a new DDG for each pass of the ACO scheduler we would have to move the code that calls ACO up to the wrapper as opposed to SchedRegion::FindOptimalSchedule.  This could be done but it would add further bloat to ScheduleDAGOptSched::schedule and would necessitate further rewrites of SchedRegion(e.g.: sched region calls the heuristic scheduler so it would also have to be moved up to the Wrapper if we wanted to have heuristic pass -> ACO pass).  I think this also answers your other comment.

Reply:
I originally looked into constructing a new DDG for each ACO pass, but upon discovering the considerations listed in the previous comment I decided to use the GetCost()/GetSpillCost() functions.

Reply:
The way the current two-pass works should already accomplish this. A new DDG is created for each pass. Using the correct DDG (all instructions 1 latency) for the first pass should make the problem much simpler for ACO. Additionally, one of the reasons why we wanted to separate the heuristic and the ACO scheduler was to use the current two-pass with ACO without much modification to ACO. ACO only has to be aware of the best minimal RP schedule found in the first pass (with stalls inserted to account for latency) and that is accomplished by [setInitialSchedule()](https://github.com/CSUS-LLVM/OptSched/blob/master/lib/Scheduler/aco.cpp#L440). A two-pass of ACO should already work with the current two-pass if both are enabled and the heuristic and B&B schedulers are disabled.

Reply:
I have previously heard that we ultimately want to run both ACO and B&B when the scheduler executes.  I thought that if we are using two pass ACO we would want to run both ACO passes first and then go on to the B&B passes, whereas the current implementation.  We should discuss the status of this commit at the meeting tomorrow especially given that two pass ACO did not give performance improvements

Reply:
If you don't rebuild the DDG you can't do things like graph transformations.

I think we need to do some refactoring to sched_region and the wrapper. There should be target hooks for all these things instead of custom modules like GCNOptSched.

We should define a scheduling pass interface that returns some information about the result of that scheduling pass. The current layout is getting untenable.

We should not have things like 1 pass scheduling, 2 pass scheduling, ACO 2 pass scheduling. There needs to be scheduling passes which can be applied independently, and in any order. They should return structs with information that has the state after scheduling.

The scheduler code itself should use inheritance to contain the state about what kind of a scheduling pass it is. I.e. there should be two versions of performSchedPass, updatePheromone based on the type of ACO scheduler is currently being run. The same should be true for 2-pass B&B. We need multiple version of sched_region.

I don't have time to work on this right now there are many things I'd like to do with the project but simply don't have the time. Since what matters here is the research and not necessarily the best software engineer principals if this is cleaned up a bit in terms of formatting it's OK to me.

What do you think Vang?

Reply:
Austin, I agree we need to spend some time and decide on how to refactor our current project. I believe we also have plans to parallelize our scheduler in the future and to do that we will definitely need to refactor our code. I will discuss this with Dr. Shobaki tomorrow on what he thinks and what the priority should be on this.

Reply:
I don't think he cares at all about refactoring.

Reply:
The only issue is that there may be other tasks that take priority over refactoring the current code. Additionally, what should the scope of the refactoring be? Should we also take parallelization into account and refactor for that, or do we do that later? For example, right now I believe the cost calculation in bb_spill can be refactored and we will need to do so for parallelization. We can't calculate the cost of two schedules in parallel because bb_spill stores the states of the schedule and we only have one copy of bb_spill. Although, If he gives us the go-ahead, then I believe we should start focusing on refactoring right away and with parallelization in mind.